aten/src/TH/generic/THTensorEvenMoreMath.cpp

```diff
@@ -138,20 +136,43 @@ void THTensor_(nonzero)(THLongTensor *subscript, THTensor *tensor)
 #endif
   THLongTensor_resize2d(subscript, numel, tensor->dim());
 
+  if (! numel) {
+    return;
+  }
+
+  int64_t dimensions = tensor->dim();
+  int64_t  sizes[dimensions+1]; // +1 faster than additional condition check inside loop
+  int64_t  idx[dimensions+1];
+  int64_t *ii;
+  int64_t *ss;
+  std::fill(sizes, sizes+dimensions+1, 0);
+  std::fill(idx, idx+dimensions+1, 0);
+
+  for (i = 0; i < dimensions; ++i) {
+    sizes[dimensions - i - 1] = THTensor_sizeLegacyNoScalars(tensor, i); // reverse order important
+  }
+
   /* Second pass populates subscripts */
   subscript_data = THLongTensor_data(subscript);
   TH_TENSOR_APPLY(scalar_t, tensor,
                   if IS_NONZERO(*tensor_data) {
-                    div = 1;
-
-                    for (dim = tensor->dim() - 1; dim >= 0; dim--) {
-                      *(subscript_data + dim) = (i/div) % THTensor_sizeLegacyNoScalars(tensor, dim);
-                      div *= THTensor_sizeLegacyNoScalars(tensor, dim);
+                    ii = idx + dimensions;
+                    for (int64_t dim = dimensions - 1; dim >= 0; dim--) {
+                      --ii;
+                      *subscript_data = *ii;
```

Comment:
this looks broken for the case where subscript is not resized (passed in as the correct size) and not contiguous, although it looks like it was broken before too.

For example:
```
>>> x
tensor([[1, 0, 1],
             [0, 1, 1]], dtype=torch.uint8)

>>> torch.nonzero(x)
tensor([[0, 0],
            [0, 2],
            [1, 1],
            [1, 2]])

>>> y=torch.full((4,2,2),5, dtype=torch.long)[:,:,1]
>>> torch.nonzero(x,out=y)
tensor([[0, 0],
             [1, 1],
             [5, 5],
             [5, 5]])
```